### ISSUE TITLE
CAn you release code for generating data?

### ISSUE THREAD

**Issue by tanphan07** [qcf-568/DocTamper#74]
Thank for your repo, can you release code for generating data

**Comment by qcf-568** [comment]
Yes, the code is under examination of the cooperative company, it can be released before the end of this year.

**Comment by tanphan07** [comment]
Thank you for your reply, I have another question. When I infer the model using .png image, I have this error: "Improper call to JPEG library in state 201". Can you give me a solution?

**Comment by qcf-568** [comment]
Please save the png image to jpeg format with quality 100 first.
code:
img = cv2.imread('your image.png')
cv2.imwrite('your image.jpg', img, [1, 100])

**Comment by qcf-568** [comment]
Please also note that this model can only process photographed images. If the image was originally jpeg and was saved as png after tampering, this model can work. However, if the image was originally png, this model cannot work.

**Comment by tanphan07** [comment]
Do you have a solution for the original png image?

**Comment by qcf-568** [comment]
This model is based on Block Artifact Grids (BAGs), which only exist in compressed images such as jpeg images. If an image is PNG before tampering, then the tampering operation will not leave anomaly trace on the BAGs, so that the model can not rely on BAGs to perform forensics. 
If you forge the photographed document, and screen capture and save the result as PNG, the frequency branch of the model can still work. 
If you forge the original PNG document such as PDF, the frequency branch of the model cannot work. In this case, is recommended to use [pure vision det model](https://github.com/qcf-568/Det_Baseline) or [pure vision seg model](https://github.com/qcf-568/MIML).
It's also notable that pure vision forensic model have significant worse cross domain generalization than frequency domain model (like this model), which is emphasized in the paper.

**Comment by tanphan07** [comment]
So, it is very important that the original image must be in jpeg format. I think in real applications, it is easy to pass this model by use the png original image, then forge image and save in png format, is it right?

**Comment by qcf-568** [comment]
In real applications, we do have a classifier at the start of the forensics system to determine whether to use frequency model or pure vision model for each input image, the png original image will be classified into the pure vision model branch. Some photographed images will be fed into the frequency model branch for better performance.

**Comment by qcf-568** [comment]
For png image,
The frequency model is good at "original jpeg image and saved as any format (including png) after tampering",
The pure vision model is good at "original png image and saved as any format after tampering".

**Comment by tanphan07** [comment]
Thank you very much for your reply, I have another question, in the paper, you don't mention how to generate tampered documents using the generation method. How can we determine the font of the character and generate another character with the same font?

**Comment by qcf-568** [comment]
Actually, for each document type, I just manually select some digital fonts that are commonly used and similar to the target font. It's hard to perform font classification and choose an accurate font automatically with deep models.

**Comment by tanphan07** [comment]
In the paper, how do you group text with similar font info (like height, width, bg color, fg color)? I suppose you use the difference of 2 characters like if different between height < error height, it has the same height, similar to the width. But for the bg color and fg color, what are the criteria for grouping?

**Comment by qcf-568** [comment]
For each text, we use [for loop] to check if every other text can meets the thresholds for grouping.
Different types of document and text have different thresholds, here is an example group of thresholds for characters in contract photos:
1. H2/1.1<=H1<=H2*1.1, H2 is the height for the compared text and H1 is the height of the group center text.
2. W2/1.1<=W1<=W2*1.1, W1 is the width of the group center text.
3. BM2-8<=BM1<=BM2+8, BM1 is the background mean of the group center text. BM1=image_region[!mask].mean()
4. FM2-8<=FM1<=FM2+8, FM1 is the foreground mean of the group center text. BM2=image_region[mask].mean()
5. BS2-4<=BS1<=BS2+4, BS1 is the background std of the group center text. BS1=image_region[!mask].std()
6. FS2-4<=FS1<=FS2+4, FS1 is the foreground std of the group center text. BS2=image_region[mask].std()
BM, BS, FM, FS are computed for each channel.

That is, the criteria for grouping color is based on the mean and std of the background and foreground respectively.

**Comment by qcf-568** [comment]
The usage of std can also controls the texture and style for background and foreground.

**Comment by tanphan07** [comment]
Thank you for your reply, but how can you figure out the group center text?

**Comment by tanphan07** [comment]
Thank you very much, hope you release code gen soon.

**Comment by tanphan07** [comment]
I have read the code dtd.py, and in this file, I saw the functions self.swin, self.vph is initialized by self.vph = torch.load('vph_imagenet.pt'), self.swin = torch.load('swin_imagenet.pt'), but I can not identify the true architecture of 2 function?. I have read 'swin_imagenet.pt' file and found that this model is a module list, but the model SwinTransformerV2 just a module

**Comment by qcf-568** [comment]
The true architecture is in the swins.py

**Comment by tanphan07** [comment]
Could you tell me the true model for self.swin, because when I print self.swin architecture, it is not the same architecture with SwinTransformerV2

**Comment by qcf-568** [comment]
The first stage of SwinV2 is dropped, the other parts keep the same.

**Comment by tanphan07** [comment]
Thank you, and i have tested performance model and realized that if I use resize operation, the model does not work well, can you handle this case ?

**Comment by qcf-568** [comment]
Please see #53

**Comment by tanphan07** [comment]
I am training model to reproduce result, but it's has a problem, iou of tampered area is decrease 
![a](https://github.com/user-attachments/assets/5bf1130e-936e-4033-850c-cc05809855a1)
And It is my dataset, can you check it?
import os
import cv2
import torch
import jpegio
import tempfile
import numpy as np
import pickle
import six
from PIL import Image
from torch.utils.data import Dataset
from albumentations.pytorch import ToTensorV2
import torchvision
from glob import glob
from io import BytesIO
from jpeg2dct.numpy import load


class TamperDataset(Dataset):
    def __init__(self, folders, total_epoch, batch_size, minq=75, maxq=100, num_compress=[1, 2, 3], total_step=100000):
        self.img_paths = []
        self.mask_paths = []
        for root in folders:
            image_folder = os.path.join(root, 'images')
            files = glob(os.path.join(image_folder, '*'))
            for f in files:
                self.img_paths.append(f)
                self.mask_paths.append(f.replace('images', 'masks').replace('jpg', 'png'))
        self.nSamples = len(self.img_paths)
        self.max_nums=self.nSamples
        # self.mode = mode
        with open('qt_table.pk','rb') as fpk:
            pks = pickle.load(fpk)
        self.pks = {}
        for k,v in pks.items():
            self.pks[k] = torch.LongTensor(v)
        self.iter = 0
        self.total_epoch = total_epoch
        self.batch_size = batch_size
        self.total_step = total_step
        self.increment = (maxq-minq)/self.total_step
        self.minq = minq
        self.maxq = maxq
        self.num_compress = num_compress
        self.epoch = 0
        self.totsr = ToTensorV2()
        self.toctsr =torchvision.transforms.Compose([torchvision.transforms.ToTensor(),torchvision.transforms.Normalize(mean=(0.485, 0.455, 0.406), std=(0.229, 0.224, 0.225))])
        self.iter = 0

    def __len__(self):
        return self.max_nums
    
    def set_epoch(self, epoch):
        self.epoch = epoch

    def __getitem__(self, index):
        
        im = Image.open(self.img_paths[index])
        mask = Image.open(self.mask_paths[index])
        
        quality = self.minq + self.increment * self.iter
        self.iter += 1
        quality = min(quality, self.maxq)
        chosen_q = []
        compressed_time = np.random.choice(self.num_compress)
        for i in range(compressed_time):
            chosen_quality = int(np.random.randint(quality, self.maxq+1))
            chosen_q.append(chosen_quality)
        
        # print(self.epoch, iter)

        mask = np.array(mask)
        mask[mask>0] = 1
    
        H,W = mask.shape
        
        mask = self.totsr(image=mask.copy())['image']
        with tempfile.NamedTemporaryFile(delete=True) as tmp:
            im = im.convert("L")
            for q in chosen_q:
                im.save(tmp,"JPEG",quality=q)
                tmp.seek(0)  # Go back to the beginning of the BytesIO buffer

                im = Image.open(tmp)
                
            # jpg = jpegio.read(tmp.name)
            # dct = jpg.coef_arrays[0].copy()

            dct = load(tmp.name,False)[0]
            h,w = dct.shape[:2]
            dct = dct.reshape(h,w,8,8).transpose(0,2,1,3).reshape(h * 8, w * 8).astype(np.int64)

            use_qtb = self.pks[chosen_q[-1]]
            im = im.convert('RGB')

        return {
            'image': self.toctsr(im),
            'label': mask.long(),
            'rgb': np.clip(np.abs(dct),0,20),
            'q':use_qtb,
            'i':q
        }

**Comment by qcf-568** [comment]
Train the model on which dataset?

**Comment by tanphan07** [comment]
I have trained in DocTamperV1-TrainingSet_new dataset

**Comment by tanphan07** [comment]
I also use Lovasz loss and SoftCrossemtropyLoss based on the paper

**Comment by qcf-568** [comment]
Do you have the official training code of this paper?

**Comment by tanphan07** [comment]
No, i have trained the model by myself to reproduce this paper

**Comment by qcf-568** [comment]
You can apply for the code via your edu email.

**Comment by tanphan07** [comment]
Can you send this code for me via hainn.B21CN320@stu.ptit.edu.vn

**Comment by TongMeng-AI** [comment]
After reading the above content, I have a question：
For original images in PNG format, the BAGs module does not work. The solution you provided is to use a pure vision detection or segmentation model. In that case, given a tampered image as input, how can we know the format of the original image?
